File: Plain Craft Launcher 2/Controls/MyImage.vb
Imports System.Net.Http
Imports PCL.Core.IO.Net.Http.Client.Request
Imports PCL.Core.Utils
Imports PCL.Core.Utils.Exts

Public Class MyImage
    Inherits Image

#Region "公开属性"

    ''' <summary>
    ''' 网络图片的缓存有效期。
    ''' 在这个时间后，才会重新尝试下载图片。
    ''' </summary>
    Public FileCacheExpiredTime As TimeSpan = TimeSpan.FromDays(14)

    ''' <summary>
    ''' 是否允许将网络图片存储到本地用作缓存。
    ''' </summary>
    Public Property EnableCache As Boolean
        Get
            Return GetValue(EnableCacheProperty)
        End Get
        Set(value As Boolean)
            SetValue(EnableCacheProperty, value)
        End Set
    End Property
    Public Shared Shadows ReadOnly EnableCacheProperty As DependencyProperty = DependencyProperty.Register(
        "EnableCache", GetType(Boolean), GetType(MyImage), New PropertyMetadata(True))

    ''' <summary>
    ''' 与 Image 的 Source 类似。
    ''' 若输入以 http 开头的字符串，则会尝试下载图片然后显示，图片会保存为本地缓存。
    ''' 支持 WebP 格式的图片。
    ''' </summary>
    Public Shadows Property Source As String '覆写 Image 的 Source 属性
        Get
            Return _Source
        End Get
        Set(value As String)
            If value = "" Then value = Nothing
            If _Source = value Then Return
            _Source = value
            If Not IsInitialized Then Return '属性读取顺序修正：在完成 XAML 属性读取后再触发图片加载（#4868）
            Load()
        End Set
    End Property
    Private _Source As String = ""
    Public Shared Shadows ReadOnly SourceProperty As DependencyProperty = DependencyProperty.Register(
        "Source", GetType(String), GetType(MyImage), New PropertyMetadata(New PropertyChangedCallback(
    Sub(sender, e) If sender IsNot Nothing Then CType(sender, MyImage).Source = e.NewValue.ToString())))

    ''' <summary>
    ''' 当 Source 首次下载失败时，会从该备用地址加载图片。
    ''' </summary>
    Public Property FallbackSource As String
        Get
            Return _FallbackSource
        End Get
        Set(value As String)
            _FallbackSource = value
        End Set
    End Property
    Private _FallbackSource As String = Nothing

    ''' <summary>
    ''' 正在下载网络图片时显示的本地图片。
    ''' </summary>
    Public Property LoadingSource As String
        Get
            Return _LoadingSource
        End Get
        Set(value As String)
            _LoadingSource = value
        End Set
    End Property
    Private _LoadingSource As String = "pack://application:,,,/images/Icons/NoIcon.png"

    Public Property CornerRadius As CornerRadius
        Get
            Return GetValue(CornerRadiusProperty)
        End Get
        Set(value As CornerRadius)
            SetValue(CornerRadiusProperty, value)
        End Set
    End Property
    Private Shared ReadOnly CornerRadiusProperty As DependencyProperty = DependencyProperty.Register(
        "CornerRadius",
        GetType(CornerRadius),
        GetType(MyImage),
        New FrameworkPropertyMetadata(
            New CornerRadius(-1),
            AddressOf OnCornerRadiusChanged)
        )

    Private Shared Sub OnCornerRadiusChanged(d As DependencyObject, e As DependencyPropertyChangedEventArgs)
        DirectCast(d, MyImage).UpdateClip()
    End Sub

    Private Sub UpdateClip() Handles Me.SizeChanged
        If (ActualWidth > 0 AndAlso ActualHeight > 0) AndAlso
            (CornerRadius.TopLeft >= 0 AndAlso CornerRadius.TopRight >= 0) Then
            Clip = New RectangleGeometry(
                New Rect(0, 0, ActualWidth, ActualHeight),
                CornerRadius.TopLeft,
                CornerRadius.TopRight)
        End If
    End Sub

#End Region

    ''' <summary>
    ''' 实际被呈现的图片地址。
    ''' </summary>
    Public Property ActualSource As String
        Get
            Return _ActualSource
        End Get
        Set(value As String)
            If value = "" Then value = Nothing
            If _ActualSource = value Then Return
            _ActualSource = value
            Dispatcher.BeginInvoke(Async Function() As Task
                Try
                    Dim bitmap As ImageSource = If(value Is Nothing, Nothing, Await Task.Run(Function() New MyBitmap(value))) '在这里先触发可能的文件读取，尽量避免在 UI 线程中读取文件
                    MyBase.Source = bitmap
                Catch ex As Exception
                    Log(ex, $"加载图片失败（{value}）")
                    Try
                        If value.StartsWithF(PathTemp) AndAlso File.Exists(value) Then File.Delete(value)
                    Catch 'ignored
                    End Try
                End Try
            End Function)
        End Set
    End Property
    Private _ActualSource As String = Nothing

    Private Sub Load() Handles Me.Initialized '属性读取顺序修正：在完成 XAML 属性读取后再触发图片加载（#4868）
        '空
        If Source Is Nothing Then
            ActualSource = Nothing
            Return
        End If
        '本地图片
        If Not Source.StartsWithF("http") Then
            ActualSource = Source
            Return
        End If
        '从缓存加载网络图片
        Dim Url As String = Source
        Dim TempPath As String = GetTempPath(Url)
        Dim TempFile As New FileInfo(TempPath)
        Dim EnableCache As Boolean = Me.EnableCache
        If EnableCache AndAlso TempFile.Exists Then
            ActualSource = TempPath
            If (Date.Now - TempFile.LastWriteTime) < FileCacheExpiredTime Then Return '无需刷新缓存
        End If

        Dispatcher.BeginInvoke(
            Async Function() As Task
                Try
                    '下载
                    ActualSource = LoadingSource '显示加载中图片

                    Dim resp = Await DownloadImageAsync(Url)
                    If Not String.IsNullOrEmpty(resp) Then
                        ActualSource = resp
                        Return
                    End If

                    resp = Await DownloadImageAsync(FallbackSource)
                    If Not String.IsNullOrEmpty(resp) Then
                        ActualSource = resp
                        Return
                    End If

                Catch ex As Exception
                    '更换备用地址
                    Log(ex, $"Online image get fail（source = {Url}, fallback = {FallbackSource}）", LogLevel.Developer)
                    '从缓存加载网络图片
                    TempPath = GetTempPath(Url)
                    TempFile = New FileInfo(TempPath)
                    If EnableCache AndAlso TempFile.Exists() Then
                        ActualSource = TempPath
                        If (Date.Now - TempFile.LastWriteTime) < FileCacheExpiredTime Then Return '无需刷新缓存
                    End If
                End Try
            End Function)
    End Sub
    Public Shared Function DownloadImageAsync(url As String) As Task(Of String)
        Return _downloadTasks.GetOrAdd(
            url,
            Function(key)
                Dim t = DownloadImageInternelAsync(key)
                t.ContinueWith(
                        Sub()
                            _downloadTasks.Remove(url, Nothing)
                        End Sub)
                Return t
            End Function)
    End Function
    Public Shared Function GetTempPath(Url As String) As String
        Return IO.Path.Combine(PathTemp, "Cache", "Images", $"{GetStringMD5(Url)}.png")
    End Function

    Private Shared ReadOnly _downloadTasks As New Concurrent.ConcurrentDictionary(Of String, Task(Of String))
    Private Shared Async Function DownloadImageInternelAsync(url As String) As Task(Of String)
        Dim tempPath = GetTempPath(url)
        Dim TempDownloadingPath = tempPath & RandomUtils.NextInt(0, 1000000)

        Try
            Directory.CreateDirectory(GetPathFromFullPath(tempPath)) '重新实现下载，以避免携带 Header（#5072）
            Using fs As New FileStream(TempDownloadingPath, FileMode.Create, FileAccess.ReadWrite, FileShare.Read)
                Using response = Await HttpRequest.Create(url).
                        WithHttpVersionOption(HttpVersion.Version30).
                        SendAsync(addMetedata:=False)
                    response.EnsureSuccessStatusCode()

                    Using nfs = Await response.AsStreamAsync()
                        fs.SetLength(0)
                        Await nfs.CopyToAsync(fs)
                    End Using
                End Using
            End Using

            File.Move(TempDownloadingPath, tempPath, True)
            Return tempPath
        Catch ex As Exception
            If File.Exists(tempPath) Then File.Delete(tempPath)
            If File.Exists(TempDownloadingPath) Then File.Delete(TempDownloadingPath)

            Log(ex, $"Try to get online image fail (url = {url}, dest = {tempPath})")
            Return String.Empty
        End Try
    End Function


End Class

File: PCL-CE/Plain Craft Launcher 2/Controls/MyImage.cs
using System.Collections.Concurrent;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using PCL.Core.Utils;
using PCL.Core.IO.Net.Http;

namespace PCL;

public class MyImage : Image
{
    public MyImage()
    {
        Initialized += (_, _) => Load();
        SizeChanged += (_, _) => UpdateClip();
    }

    /// <summary>
    ///     实际被呈现的图片地址。
    /// </summary>
    public string ActualSource
    {
        get => field;
        set
        {
            if (string.IsNullOrEmpty(value))
                value = null;
            if ((field ?? "") == (value ?? ""))
                return;
            field = value;
            Dispatcher.BeginInvoke(new Func<Task>(async () =>
            {
                try
                {
                    ImageSource bitmap = value is null ? null : await Task.Run(() => new MyBitmap(value));
                    base.Source = bitmap;
                }
                catch (Exception ex)
                {
                    ModBase.Log(ex, $"加载图片失败（{value}）");
                    try
                    {
                        if (value.StartsWithF(ModBase.pathTemp) && File.Exists(value)) File.Delete(value);
                    }
                    catch
                    {
                    }
                }
            })); // 在这里先触发可能的文件读取，尽量避免在 UI 线程中读取文件
            // ignored
        }
    }

    private void Load() // 属性读取顺序修正：在完成 XAML 属性读取后再触发图片加载（#4868）
    {
        if (Source is null)
        {
            ActualSource = null;
            return;
        }

        if (!Source.StartsWithF("http"))
        {
            ActualSource = Source;
            return;
        }

        var url = Source;
        var tempPath = GetTempPath(url);
        var tempFile = new FileInfo(tempPath);
        var enableCache = this.EnableCache;
        if (enableCache && tempFile.Exists)
        {
            ActualSource = tempPath;
            if (DateTime.Now - tempFile.LastWriteTime < fileCacheExpiredTime)
                return; // 无需刷新缓存
        }

        Dispatcher.BeginInvoke(new Func<Task>(async () =>
        {
            try
            {
                // 下载
                ActualSource = LoadingSource;

                var resp = await DownloadImageAsync(url);
                if (!string.IsNullOrEmpty(resp))
                {
                    ActualSource = resp;
                    return;
                }

                resp = await DownloadImageAsync(FallbackSource);
                if (!string.IsNullOrEmpty(resp))
                {
                    ActualSource = resp;
                    return;
                }
            }
            catch (Exception ex)
            {
                // 更换备用地址
                ModBase.Log(ex, $"Online image get fail（source = {url}, fallback = {FallbackSource}）", ModBase.LogLevel.Developer);
                tempPath = GetTempPath(url);
                tempFile = new FileInfo(tempPath);
                if (enableCache && tempFile.Exists)
                {
                    ActualSource = tempPath;
                    if (DateTime.Now - tempFile.LastWriteTime < fileCacheExpiredTime)
                        return;
                }
            }
        }));
    }

    public static Task<string> DownloadImageAsync(string url)
    {
        return _downloadTasks.GetOrAdd(url, key =>
        {
            var t = DownloadImageInternalAsync(key);
            t.ContinueWith(_ => _downloadTasks.TryRemove(url, out _));
            return t;
        });
    }

    public static string GetTempPath(string url)
    {
        return Path.Combine(ModBase.pathTemp, "Cache", "Images", $"{ModBase.GetStringMD5(url)}.png");
    }

    private static readonly ConcurrentDictionary<string, Task<string>> _downloadTasks = new();

    private static async Task<string> DownloadImageInternalAsync(string url)
    {
        var tempPath = GetTempPath(url);
        var tempDownloadingPath = tempPath + RandomUtils.NextInt(0, 1000000);

        try
        {
            Directory.CreateDirectory(ModBase.GetPathFromFullPath(tempPath)); // 重新实现下载，以避免携带 Header（#5072）
            using (var fs = new FileStream(tempDownloadingPath, FileMode.Create, FileAccess.ReadWrite, FileShare.Read))
            {
                using (var response = await HttpRequest.Create(url)
                           .WithHttpVersionOption(HttpVersion.Version30)
                           .SendAsync(addMetedata: false))
                {
                    response.EnsureSuccessStatusCode();

                    using (var nfs = await response.AsStreamAsync())
                    {
                        fs.SetLength(0L);
                        await nfs.CopyToAsync(fs);
                    }
                }
            }

            File.Move(tempDownloadingPath, tempPath, true);
            return tempPath;
        }
        catch (Exception ex)
        {
            if (File.Exists(tempPath)) File.Delete(tempPath);
            if (File.Exists(tempDownloadingPath)) File.Delete(tempDownloadingPath);

            ModBase.Log(ex, $"Try to get online image fail (url = {url}, dest = {tempPath})");
            return string.Empty;
        }
    }

    #region 公开属性

    public TimeSpan fileCacheExpiredTime = TimeSpan.FromDays(14d);

    public bool EnableCache
    {
        get => (bool)GetValue(EnableCacheProperty);
        set => SetValue(EnableCacheProperty, value);
    }

    public new static readonly DependencyProperty EnableCacheProperty =
        DependencyProperty.Register("EnableCache", typeof(bool), typeof(MyImage), new PropertyMetadata(true));

    /// <summary>
    ///     与 Image 的 Source 类似。
    ///     若输入以 http 开头的字符串，则会尝试下载图片然后显示，图片会保存为本地缓存。
    ///     支持 WebP 格式的图片。
    /// </summary>
    public new string Source // 覆写 Image 的 Source 属性
    {
        get => field;
        set
        {
            if (string.IsNullOrEmpty(value))
                value = null;
            if ((field ?? "") == (value ?? ""))
                return;
            field = value;
            if (!IsInitialized)
                return; // 属性读取顺序修正：在完成 XAML 属性读取后再触发图片加载（#4868）
            Load();
        }
    } = "";

    public new static readonly DependencyProperty SourceProperty = DependencyProperty.Register("Source", typeof(string),
        typeof(MyImage), new PropertyMetadata((sender, e) =>
        {
            if (sender is not null) ((MyImage)sender).Source = e.NewValue.ToString();
        }));

    /// <summary>
    ///     当 Source 首次下载失败时，会从该备用地址加载图片。
    /// </summary>
    public string FallbackSource { get; set; }

    /// <summary>
    ///     正在下载网络图片时显示的本地图片。
    /// </summary>
    public string LoadingSource { get; set; } = "pack://application:,,,/images/Icons/NoIcon.png";
    public CornerRadius CornerRadius
    {
        get => (CornerRadius)GetValue(CornerRadiusProperty);
        set => SetValue(CornerRadiusProperty, value);
    }
    private static readonly DependencyProperty CornerRadiusProperty =
        DependencyProperty.Register(
            "CornerRadius",
            typeof(CornerRadius),
            typeof(MyImage),
            new FrameworkPropertyMetadata(
                new CornerRadius(-1),
                OnCornerRadiusChanged)
        );

    private static void OnCornerRadiusChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
    {
        ((MyImage)d).UpdateClip();
    }

    private void UpdateClip() // Handles Me.SizeChanged will be added separately
    {
        if (ActualWidth > 0 && ActualHeight > 0 &&
            CornerRadius.TopLeft >= 0 && CornerRadius.TopRight >= 0)
        {
            Clip = new RectangleGeometry(
                new Rect(0, 0, ActualWidth, ActualHeight),
                CornerRadius.TopLeft,
                CornerRadius.TopRight);
        }
    }
    #endregion
}